src/gs-update-monitor.c

```diff
@@ -48,10 +52,46 @@ struct _GsUpdateMonitor {
 	guint		 check_hourly_id;		/* and then every hour */
 	guint		 check_daily_id;		/* every 3rd day */
 	guint		 notification_blocked_id;	/* rate limit notifications */
+
+	MwscScheduler	*scheduler;
+	GHashTable	*scheduled_updates; /* (element-type utf8 UpdateScheduleHelper) */
+	GCancellable	*scheduled_updates_cancellable;
 };
 
 G_DEFINE_TYPE (GsUpdateMonitor, gs_update_monitor, G_TYPE_OBJECT)
 
+typedef struct {
+	GsUpdateMonitor *monitor;
+	MwscScheduleEntry *entry;
+	GsApp *app;
+	gulong download_now_handler_id;
+	gulong invalidate_handler_id;
+} UpdateScheduleHelper;
+
+static void
+download_schedule_helper_free (UpdateScheduleHelper *helper)
+{
+	if (helper->entry != NULL) {
+		g_debug ("Unscheduling update for app %s, with entry id %s",
+			 gs_app_get_unique_id (helper->app),
+			 mwsc_schedule_entry_get_id (helper->entry));
+
+		if (helper->download_now_handler_id > 0)
+			g_signal_handler_disconnect (helper->entry,
+						     helper->download_now_handler_id);
+		if (helper->invalidate_handler_id > 0)
+			g_signal_handler_disconnect (helper->entry,
+						     helper->invalidate_handler_id);
+		mwsc_schedule_entry_remove_async (helper->entry, NULL, NULL, NULL);
+		g_clear_object (&helper->entry);
+	}
+
+	g_object_unref (helper->app);
+	g_free (helper);
```

Comment:
Remember to use `g_slice_free` if you have moved to `g_slice_new0` below.

Reply:
@mariospr See my other reply to the `g_slice_new0`.

Reply:
Ok